core/thread/test/testRWLock.cxx

```diff
@@ -0,0 +1,398 @@
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      m->WriteLock();
+   }
+}
+
+template <typename M>
+void testReadLock(M *m, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      m->ReadLock();
+   }
+}
+
+template <typename M>
+void testNonReentrantLock(M *m, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      m->lock();
+      m->unlock();
+   }
+}
+
+template <typename M>
+void testWriteTUnLock(M *m, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      m->UnLock();
+   }
+}
+
+template <typename M>
+void testWriteUnLock(M *m, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      m->WriteUnLock();
+   }
+}
+
+template <typename M>
+void testReadUnLock(M *m, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      m->ReadUnLock();
+   }
+}
+
+void testWriteGuard(TVirtualMutex *m, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      TLockGuard guard(m);
+   }
+}
+
+void testReadGuard(TVirtualRWMutex *m, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      m->ReadLock();
+      m->ReadUnLock();
+   }
+}
+
+struct Globals {
+   size_t fFirst = 0;
+   size_t fSecond = 0;
+   size_t fThird = 0;
+};
+
+void writer(TVirtualRWMutex *m, Globals *global, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      {
+         TLockGuard guard(m);
+         global->fFirst++;
+         // Waste some time
+         for (size_t k = 0; k < 100; ++k) {
+            global->fSecond += global->fThird * global->fFirst + k;
+         }
+         global->fThird++;
+      }
+      gSystem->Sleep(3 /* milliseconds */); // give sometimes to the readers
+   }
+}
+
+void reader(TVirtualRWMutex *m, Globals *global, size_t repetition)
+{
+   for (size_t i = 0; i < repetition; ++i) {
+      m->ReadLock();
+      ASSERT_EQ(global->fFirst, global->fThird);
+      m->ReadUnLock();
+      gSystem->Sleep(1 /* milliseconds */); // give sometimes to the writers
+   }
+}
+
+void concurrent(TVirtualRWMutex *m, size_t nwriters, size_t nreaders, size_t repetition)
+{
+   // ROOT::EnableThreadSafety();
+
+   std::vector<std::thread> threads;
+
+   Globals global;
+
+   for (size_t i = 0; i < nwriters; ++i) {
+      threads.push_back(std::thread([&]() { writer(m, &global, repetition); }));
+   }
+   for (size_t i = 0; i < nreaders; ++i) {
+      threads.push_back(std::thread([&]() { reader(m, &global, repetition); }));
+   }
+
+   for (auto &&th : threads) {
+      th.join();
+   }
+}
+
+template <typename T>
+void Reentrant(T &m)
+{
+
+   m.ReadLock();
+   m.ReadLock();
+   m.ReadLock();
+
+   m.WriteLock();
+
+   m.ReadLock();
+   m.ReadLock();
+
+   m.WriteLock();
+
+   m.ReadLock();
+
+   m.ReadUnLock();
+   m.WriteUnLock();
+   m.ReadUnLock();
+   m.ReadUnLock();
+   m.WriteUnLock();
+   m.ReadUnLock();
+   m.ReadUnLock();
+   m.ReadUnLock();
+}
+
+constexpr size_t gRepetition = 10000000;
```

Comment:
The other day we have had a discussion about performance benchmarking on a different use-cases. It turns out there is a framework, very similar to `googletest` called `googlebenchmark`. It saves a lot of time writing this boilerplate code. We were considering on introducing this in a PR.